common/tqname/tqname.go

```diff
@@ -0,0 +1,196 @@
+	baseName := name
+	partition := 0
+	versionSet := ""
+
+	if strings.HasPrefix(name, mangledTaskQueuePrefix) {
+		suffixOff := strings.LastIndex(name, suffixDelimiter)
+		if suffixOff <= len(mangledTaskQueuePrefix) {
+			return Name{}, fmt.Errorf("invalid task queue name %q", name)
+		}
+		baseName = name[len(mangledTaskQueuePrefix):suffixOff]
+
+		suffix := name[suffixOff+1:]
+		if partitionOff := strings.LastIndex(suffix, versionSetDelimiter); partitionOff == 0 {
+			return Name{}, fmt.Errorf("invalid task queue name %q", name)
+		} else if partitionOff > 0 {
+			// pull out version set
+			versionSet, suffix = suffix[:partitionOff], suffix[partitionOff+1:]
+		}
+
+		var err error
+		partition, err = strconv.Atoi(suffix)
+		if err != nil || partition < 0 || partition == 0 && len(versionSet) == 0 {
+			return Name{}, fmt.Errorf("invalid task queue name %q", name)
+		}
+	}
+
+	return Name{
+		baseName:   baseName,
+		partition:  partition,
+		versionSet: versionSet,
+	}, nil
+}
+
+// FromBaseName takes a base name and returns a Name. Returns an error if name looks like a
+// mangled name.
+func FromBaseName(name string) (Name, error) {
+	if strings.HasPrefix(name, mangledTaskQueuePrefix) {
+		return Name{}, fmt.Errorf("base name %q must not have prefix %q", name, mangledTaskQueuePrefix)
+	}
+	return Name{baseName: name}, nil
+}
+
+// MustFromBaseName is a convenience function for tests, when name is guaranteed to be a base name.
+func MustFromBaseName(name string) Name {
+	n, err := FromBaseName(name)
+	if err != nil {
+		return InvalidName
+	}
+	return n
+}
+
+// IsRoot returns true if this task queue is a root partition.
+func (tn Name) IsRoot() bool {
+	return tn.partition == 0
+}
+
+// WithPartition returns a new Name with the same base and version set but a different partition.
+func (tn Name) WithPartition(partition int) Name {
+	nn := tn
+	nn.partition = partition
+	return nn
+}
+
+// Root is shorthand for WithPartition(0).
+func (tn Name) Root() Name {
+	return tn.WithPartition(0)
+}
+
+// WithVersionSet returns a new Name with the same base and partition but a different version set.
+func (tn Name) WithVersionSet(versionSet string) Name {
+	nn := tn
+	nn.versionSet = versionSet
+	return nn
+}
+
+// BaseNameString returns the base name for a task queue. This should be used when looking up
+// settings in dynamic config, and pretty much nowhere else. To get the name of the root
+// partition, use tn.Root().FullName().
+func (tn Name) BaseNameString() string {
+	return tn.baseName
+}
+
+// Partition returns the partition number for a task queue.
+func (tn Name) Partition() int {
+	return tn.partition
+}
+
+// VersionSet returns the version set for a task queue.
+func (tn Name) VersionSet() string {
+	return tn.versionSet
+}
+
+// Parent returns a Name for the parent partition, using the given branching degree.
+func (tn Name) Parent(degree int) Name {
+	if tn.IsRoot() || degree == 0 {
+		return InvalidName
```

Comment:
I don't see any code checking against the `InvalidName` sentinel _value_, and it's not obvious from the API that it should be checked. So I think `(Name, bool)` or `(Name, error)` with a sentinel _error_ are better return types here. I also think we should get rid of `InvalidName` because this is the only non-test use case, and we can panic in other cases because they're initiated by tests.